test/data/s3/s3_data.py

```diff
@@ -0,0 +1,451 @@
+            b"bytes: \x00\x01\x02",
+            "just a string"]
+    expected = {}
+    objs = []
+    for text in data:
+        blob = to_bytes(text)
+        checksum = sha1(blob).hexdigest()
+        key = str(uuid4())
+        expected[os.path.join(put_prefix, key)] = {None: ExpectedResult(
+            size=len(blob),
+            checksum=checksum,
+            content_type=None,
+            metadata=None,
+            range=None)}
+        objs.append((key, text))
+        if meta is not None:
+            for content_type, usermeta in meta.values():
+                key = str(uuid4())
+                expected[os.path.join(put_prefix, key)] = {None: ExpectedResult(
+                    size=len(blob),
+                    checksum=checksum,
+                    content_type=content_type,
+                    metadata=usermeta,
+                    range=None)}
+                objs.append(S3PutObject(
+                    key=key,
+                    value=text,
+                    content_type=content_type,
+                    metadata=usermeta))
+    return {'argvalues': [(put_prefix, objs, expected)],
+            'ids': ['put_strings']}
+
+def pytest_put_blobs_case(meta=None):
+    put_prefix = os.path.join(S3ROOT, PUT_PREFIX)
+    cases = []
+    ids = []
+    for prefix, filespecs in BIG_DATA:
+        expected = {}
+        blobs = []
+        for fname, size in filespecs.items():
+            blob = RandomFile(prefix, fname, size)
+            checksum = blob.checksum()
+            key = str(uuid4())
+            expected[os.path.join(put_prefix, key)] = {None: ExpectedResult(
+                size=blob.size,
+                checksum=checksum,
+                content_type=None,
+                metadata=None,
+                range=None)}
+            blobs.append((key, blob.data))
+            if meta is not None:
+                for content_type, usermeta in meta.values():
+                    key = str(uuid4())
+                    expected[os.path.join(put_prefix, key)] = {None: ExpectedResult(
+                        size=len(blob),
+                        checksum=checksum,
+                        content_type=content_type,
+                        metadata=usermeta,
+                        range=None)}
+                    blobs.append(S3PutObject(
+                        key=key,
+                        value=blob.data,
+                        content_type=content_type,
+                        metadata=usermeta))
+        ids.append(prefix)
+        cases.append((put_prefix, blobs, expected))
+    return {'argvalues': cases, 'ids': ids}
+
+def ensure_test_data():
+    # update S3ROOT in __init__.py to get a fresh set of data
+    print('Ensuring that test data exists at %s' % S3ROOT)
+    mark = urlparse(os.path.join(S3ROOT, 'ALL_OK'))
+    try:
+        # Check if parquet data exists and has been modified in the last
+        # 29 days (default TTL - 1).
```

Comment:
What is "default TTL" here?

Reply:
Also, roughly how long does it take to generate test data? If it is fast enough I'd err on the side of having shorter TTLs

Reply:
It seems to take under 5 min from an EC2 instance. YMMV (a lot of files and some large ones are uploaded). I don't know if there is a lot of value in re-uploading a lot though so I think 29 days is enough. I'll remove the reference to default TTL; this is probably something Netflix specific.

Reply:
I'd maybe go lower, like a day or two maybe? I mean, it is probably fine as is, but generally when I see a code path that gets executed only _once every X weeks_ it makes me a little uneasy, as I worry that if there is an obscure bug by some chance, it'll take a while to spot

Reply:
I left as is for now. I did clarify the TTL requirement though. IMHO, uploading data can be expensive so I didn't think it very useful to re-upload too much. But don't have a strong opinion.